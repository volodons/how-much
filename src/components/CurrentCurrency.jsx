import { useDispatch, useSelector } from 'react-redux';
import Autocomplete from '@mui/material/Autocomplete';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';

import {
    StyledBox,
    StyledTypography,
    StyledTypographyButton,
    StyledTextField,
    StyledIconButton,
    StyledPriceChangeIcon,
} from '../styled/styledCurrentCurrency';
import { currenciesActions } from '../redux/ducks/currenciesDuck';

const CurrentCurrency = () => {
    const dispatch = useDispatch();
    const { allCurrencies } = useSelector((state) => state.currencies);

    const handleChangeBaseCurrency = (currency) => {
        if (currency) {
            dispatch({
                type: currenciesActions.SET_BASE_CURRENCY.type,
                payload: currency.code,
            });
        } else {
            dispatch({
                type: currenciesActions.SET_BASE_CURRENCY.type,
                payload: '',
            });
        }
    };

    return (
        <StyledBox>
            <StyledTypography variant="h2">Current Currency</StyledTypography>
            <Formik
                initialValues={{ baseCurrency: '' }}
                onSubmit={(values) => {
                    handleChangeBaseCurrency(values.baseCurrency);
                }}
                validationSchema={Yup.object().shape({
                    baseCurrency: Yup.object().required(
                        'This field is required'
                    ),
                })}
            >
                <Form>
                    <Grid item>
                        <Field name="baseCurrency">
                            {({ field, form }) => (
                                <Autocomplete
                                    {...field}
                                    options={allCurrencies}
                                    getOptionLabel={(option) =>
                                        option.code || ''
                                    }
                                    renderInput={(params) => (
                                        <StyledTextField
                                            {...params}
                                            label="Choose Your Currency"
                                        />
                                    )}
                                    renderOption={(props, option) => (
                                        <Box component="li" {...props}>
                                            {option.code}
                                        </Box>
                                    )}
                                    onChange={(event, value) => {
                                        form.setFieldValue(
                                            'baseCurrency',
                                            value
                                        );
                                    }}
                                />
                            )}
                        </Field>
                        <ErrorMessage name="baseCurrency" />
                    </Grid>
                    <StyledIconButton type="submit">
                        <StyledPriceChangeIcon />
                        <StyledTypographyButton variant="button">
                            Set
                        </StyledTypographyButton>
                    </StyledIconButton>
                </Form>
            </Formik>
        </StyledBox>
    );
};

export default CurrentCurrency;
