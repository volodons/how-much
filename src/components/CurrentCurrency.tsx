import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Autocomplete from '@mui/material/Autocomplete';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import { Formik, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';

import { RootState } from '../redux';
import { currenciesActions } from '../redux/ducks/currenciesDuck';
import {
    StyledBox,
    StyledTypography,
    StyledTypographyButton,
    StyledForm,
    StyledTextField,
    StyledIconButton,
    StyledPriceChangeIcon,
} from '../styled/styledCurrentCurrency';

interface Currency {
    id: number;
    code: string;
    exchangeRate: number;
    featured: boolean;
}

interface Values {
    baseCurrency: string;
}

const CurrentCurrency: React.FC = () => {
    const dispatch = useDispatch();
    const { currencies, baseCurrency } = useSelector(
        (state: RootState) => state.currencies
    );

    const handleChangeBaseCurrency = (currency: string | null) => {
        if (currency) {
            dispatch({
                type: currenciesActions.SET_BASE_CURRENCY.type,
                payload: currency,
            });
        } else {
            dispatch({
                type: currenciesActions.SET_BASE_CURRENCY.type,
                payload: '',
            });
        }
    };

    useEffect(() => {
        dispatch({ type: currenciesActions.FETCH_BASE_CURRENCY.type });
    }, []);

    const initialValues: Values = {
        baseCurrency: baseCurrency || '',
    };

    const inputOrOption = Yup.lazy((value) => {
        if (typeof value === 'string' || typeof value === 'object') {
            return Yup.mixed().required('This field is required');
        }
        return Yup.string().typeError(
            'Value must be your input or one of suggested options in menu'
        );
    });

    return (
        <StyledBox data-testid="current-currency-component">
            <StyledTypography variant="h2">Current Currency</StyledTypography>
            <Formik
                initialValues={initialValues}
                onSubmit={(values) => {
                    let selectedCurrencyCode = '';
                    if (typeof values.baseCurrency === 'string') {
                        const selectedCurrency = currencies.find(
                            (currency) => currency.code === values.baseCurrency
                        );
                        selectedCurrencyCode = selectedCurrency
                            ? selectedCurrency.code
                            : '';
                    } else {
                        selectedCurrencyCode = (values.baseCurrency as Currency)
                            .code;
                    }
                    if (selectedCurrencyCode) {
                        handleChangeBaseCurrency(selectedCurrencyCode);
                    }
                }}
                validationSchema={Yup.object().shape({
                    baseCurrency: inputOrOption,
                })}
            >
                <StyledForm>
                    <Grid item>
                        <Field name="baseCurrency">
                            {({ field, form }: { field: any; form: any }) => (
                                <Autocomplete
                                    {...field}
                                    options={currencies}
                                    getOptionLabel={(option: Currency) =>
                                        option.code || baseCurrency || ''
                                    }
                                    renderInput={(params) => (
                                        <StyledTextField
                                            {...params}
                                            label="Choose Your Currency"
                                        />
                                    )}
                                    renderOption={(props, option: Currency) => (
                                        <Box component="li" {...props}>
                                            {option.code}
                                        </Box>
                                    )}
                                    onChange={(
                                        event: React.ChangeEvent<{}>,
                                        value: Currency | string | null
                                    ) => {
                                        form.setFieldValue(
                                            'baseCurrency',
                                            value || ''
                                        );
                                    }}
                                />
                            )}
                        </Field>
                        <ErrorMessage name="baseCurrency" />
                    </Grid>
                    <StyledIconButton type="submit">
                        <StyledPriceChangeIcon />
                        <StyledTypographyButton variant="button">
                            Fetch Rates
                        </StyledTypographyButton>
                    </StyledIconButton>
                </StyledForm>
            </Formik>
        </StyledBox>
    );
};

export default CurrentCurrency;
