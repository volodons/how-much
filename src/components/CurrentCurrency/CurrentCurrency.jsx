import CURRENCIES from "../../const/currencies/currencies";
import Flag from "react-world-flags";
import Autocomplete from "@mui/material/Autocomplete";
import Box from "@mui/material/Box";
import { styled } from "@mui/material/styles";
import TextField from "@mui/material/TextField";
import Typography from "@mui/material/Typography";

const StyledBox = styled(Box)`
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 2rem;
    padding: 2rem;
    text-align: center;
    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    border-radius: 10px;
    background: linear-gradient(to right, #0390fc, #562fff);
    transition: box-shadow 0.3s ease-in-out;
    &:hover {
        box-shadow: 4px 4px 8px rgba(0, 0, 0, 0.5);
        transform: scale(1.02);
    }
    @media (max-width: 1000px) {
        width: 75%;
    }
    @media (max-width: 800px) {
        width: 90%;
    }
    @media (max-width: 600px) {
        width: auto;
    }
`;

const StyledTypography = styled(Typography)`
    margin-bottom: 2rem;
    color: #f5f7ff;
    text-decoration: underline;
    font-family: lobster, sans-serif;
    text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    &:hover {
        text-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
    }
    @media (max-width: 600px) {
        font-size: 2.5rem;
    }
`;

const StyledTextFiled = styled(TextField)`
    & .MuiOutlinedInput-root {
        &:hover fieldset {
            border-color: #f5f7ff;
        }
    }
    & .MuiAutocomplete-inputRoot {
        color: #f5f7ff;
    }
    & .MuiButtonBase-root {
        color: #f5f7ff;
    }
    & .MuiInputLabel-root {
        color: #f5f7ff;
        &:active {
            color: #f5f7ff;
        }
        &:focus {
            color: #f5f7ff;
        }
    }
    & .MuiOutlinedInput-notchedOutline {
        border-width: 2px;
        border-color: #f5f7ff;
        border-radius: 10px;
        &:active {
            border-color: #f5f7ff;
        }
        &:focus {
            border-color: #f5f7ff;
        }
    }
`;

const StyledFlag = styled(Flag)`
    width: 2rem;
    height: 2rem;
    margin-right: 1rem;
`;

const CurrentCurrency = () => {
    return (
        <StyledBox>
            <StyledTypography variant="h2">Current Currency</StyledTypography>
            <Autocomplete
                defaultValue={CURRENCIES[0].currency}
                options={CURRENCIES}
                getOptionLabel={(option) =>
                    option.currency || CURRENCIES[0].currency
                }
                renderInput={(params) => (
                    <StyledTextFiled {...params} label="Choose Your Currency" />
                )}
                renderOption={(props, option) => (
                    <Box component="li" {...props}>
                        <StyledFlag code={option.country} />
                        {option.currency}
                    </Box>
                )}
            />
        </StyledBox>
    );
};

export default CurrentCurrency;
