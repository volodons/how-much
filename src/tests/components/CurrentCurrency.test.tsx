import {
    render,
    fireEvent,
    waitFor,
    screen,
    act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import store from '../../redux';
import CurrentCurrency from '../../components/CurrentCurrency';

describe('CurrentCurrency component', () => {
    it('should be a functional component', () => {
        expect(CurrentCurrency).toBeInstanceOf(Function);
    });

    it('should render Current Currency heading', async () => {
        render(
            <Provider store={store}>
                <CurrentCurrency />
            </Provider>
        );

        await waitFor(() => {
            expect(screen.getByText('Current Currency')).toBeInTheDocument();
        });
    });

    it('should dispatch action to fetch base currency on mount', () => {
        render(
            <Provider store={store}>
                <CurrentCurrency />
            </Provider>
        );
    });

    it('should dispatch action to set base currency when autocomplete option is selected', () => {
        render(
            <Provider store={store}>
                <CurrentCurrency />
            </Provider>
        );

        const autocompleteInput = screen.getByRole('combobox');

        fireEvent.change(autocompleteInput, { target: { value: 'USD' } });
    });

    it('should dispatch action to fetch exchange rates when Fetch Rates button is clicked', () => {
        render(
            <Provider store={store}>
                <CurrentCurrency />
            </Provider>
        );

        const fetchRatesButton = screen.getByText('Fetch Rates');

        act(() => {
            fireEvent.click(fetchRatesButton);
        });
    });

    it('should match a snapshot', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <CurrentCurrency />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
