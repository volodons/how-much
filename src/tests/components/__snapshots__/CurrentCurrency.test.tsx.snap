// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentCurrency component should match a snapshot 1`] = `
<div
  className="MuiBox-root css-rbet3"
  data-testid="current-currency-component"
>
  <h2
    className="MuiTypography-root MuiTypography-h2 css-1tpbzme-MuiTypography-root"
  >
    Current Currency
  </h2>
  <form
    action="#"
    className="css-11mmnht"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        aria-owns={null}
        className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
        name="baseCurrency"
        onBlur={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1mnimye-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Choose Your Currency
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              aria-invalid={false}
              autoCapitalize="none"
              autoComplete="off"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              readOnly={false}
              required={false}
              role="combobox"
              spellCheck="false"
              type="text"
              value=""
            />
            <div
              className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Clear"
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="CloseIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </button>
              <button
                aria-label="Open"
                className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </button>
            </div>
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Choose Your Currency
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-bk4gzm-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1iq9u08-MuiSvgIcon-root"
        data-testid="PriceChangeIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 4H4c-1.11 0-1.99.89-1.99 2L2 18c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V6c0-1.11-.89-2-2-2m-8 6H8v1h3c.55 0 1 .45 1 1v3c0 .55-.45 1-1 1h-1v1H8v-1H6v-2h4v-1H7c-.55 0-1-.45-1-1V9c0-.55.45-1 1-1h1V7h2v1h2zm4 6.25-2-2h4zM14 10l2-2 2 2z"
        />
      </svg>
      <span
        className="MuiTypography-root MuiTypography-button css-xj5rr6-MuiTypography-root"
      >
        Fetch Rates
      </span>
    </button>
  </form>
</div>
`;
